ord_schema/validations.py

```diff
@@ -120,16 +127,78 @@ def validate_reaction_identifier(message):
 def validate_reaction_input(message):
     if len(message.components) == 0:
         raise ValueError('Reaction inputs must have at least one component')
+    for component in message.components:
+        if not component.WhichOneof('amount'):
+            raise ValueError('Reaction input\'s components require an amount')
     return message
 
 
+def pubchem_resolve(value_type, value):
+    return request.urlopen('https://pubchem.ncbi.nlm.nih.gov/rest/pug'
+                           f'/compound/{value_type}/{value}/property/'
+                           'IsomericSMILES/txt').read().decode().strip()
+
 def validate_compound(message):
+    # pylint: disable=too-many-branches
     if len(message.identifiers) == 0:
         raise ValueError('Compounds must have at least one identifier')
+
+    # If no structural identifiers have been used, iterate through available
+    # identifiers until we can resolve one to a SMILES string
+    if not any(identifier.type in COMPOUND_STRUCTURAL_IDENTIFIERS_ for
+               identifier in message.identifiers):
+        for identifier in message.identifiers:
+            if identifier.type == identifier.NAME:
+                try:
+                    smiles = pubchem_resolve('name', identifier.value)
+                    new_identifier = message.identifiers.add()
+                    new_identifier.type = new_identifier.SMILES
+                    new_identifier.value = smiles
+                    new_identifier.details = 'NAME resolved by PubChem'
+                    break
+                except HTTPError:
+                    pass
+
+    # Try to create an RDKit binary identifier
+    # TODO(ccoley) add more sources for RDKIT_BINARY
+    if Chem and not any(identifier.type == identifier.RDKIT_BINARY for
```

Comment:
Consider moving this branch out into a separate function to make the overall function more readable (and make for easier testing).

Reply:
Could you clarify which part you would like to see separated out? This block has a dual role: validating defined identifiers and creating the ```RDKIT_BINARY``` identifier. The former role should be inside this function at least.

Reply:
Yes, I was referring to the creation of the RDKIT_BINARY identifier.

Reply:
Since this ends up only being a few lines of code (beyond what is required to validate the identifiers), I've kept it as-is for now. When we implement more extensive canonicalization to resolve tautomers, salts, charges, etc., I will make separate functions for  ```canonicalize_mol(mol)``` and ```add_rdkit_binary(message, mol)```